### ISSUE TITLE
stride_tricks.as_strides loses custom dtype

### ISSUE THREAD

**Issue by pitrou** [numpy/numpy#9161]
```python
>>> from numpy.core.test_rational import rational
>>> from numpy.lib import stride_tricks
>>> a = np.zeros(2, dtype=np.dtype(rational))
>>> a
array([rational(0), rational(0)], dtype=rational)
>>> stride_tricks.as_strided(a, strides=a.strides)
array([[0 0 0 0 0 0 0 0], [0 0 0 0 0 0 0 0]], 
      dtype='|V8')
```

**Comment by mhvk** [comment]
Odd.  The reason seems to be that the `dtype` is not one that is understood by `__array_interface__`, which is used to pass the strides on in https://github.com/numpy/numpy/blob/master/numpy/lib/stride_tricks.py#L96; it turns it into the `V8` void dtype.

Now there is a check on a wrong void dtype already in https://github.com/numpy/numpy/blob/master/numpy/lib/stride_tricks.py#L104. Maybe the easiest would be to just remove the `if` statement, i.e., always assign `array.dtype = x.dtype` -- I don't see in what situation this would be incorrect. 

But following up on `git blame`Indeed, I see that in #5519 the `if` statement made this specific; before, any `void` would have had its dtype reset. So, cc #5519's author, @jaimefrio.

**Comment by mhvk** [comment]
Well, looking a bit more, I do see again why one cannot just assign the `dtype` -- that had another problem, #5081, which #5519 was meant to fix...

**Comment by pitrou** [comment]
Well it seems one can now assign the `dtype` correctly.

**Comment by mhvk** [comment]
fixed by #9164
